src/lib/openjp2/segmented_stream.c

```diff
@@ -0,0 +1,335 @@
+		return OPJ_FALSE;
+
+	seg = (opj_buffer_t*)opj_vec_back(&seg_buf->segments);
+	seg->owns_data = OPJ_TRUE;
+
+	return OPJ_TRUE;
+}
+
+void opj_seg_buf_incr_cur_seg_offset(opj_seg_buf_t* seg_buf, OPJ_OFF_T offset) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return;
+	cur_seg = (opj_buffer_t*)(seg_buf->segments.data[seg_buf->cur_seg_id]);
+	opj_buffer_incr_offset(cur_seg, offset);
+	if (cur_seg->offset == cur_seg->len) {
+		opj_seg_buf_increment(seg_buf);
+	}
+
+}
+
+
+/**
+* Zero copy read of contiguous chunk from current segment.
+* Returns OPJ_FALSE if unable to get a contiguous chunk, OPJ_TRUE otherwise
+*/
+OPJ_BOOL opj_seg_buf_zero_copy_read(opj_seg_buf_t* seg_buf,
+										OPJ_BYTE** ptr,
+										OPJ_SIZE_T chunk_len) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return OPJ_FALSE;
+	cur_seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id);
+
+	if (cur_seg->offset + chunk_len <= cur_seg->len) {
+		*ptr = cur_seg->buf + cur_seg->offset;
+		opj_seg_buf_read(seg_buf, NULL, chunk_len);
+		return OPJ_TRUE;
+	}
+	return OPJ_FALSE;
+}
+
+OPJ_BOOL opj_seg_buf_copy_to_contiguous_buffer(opj_seg_buf_t* seg_buf, OPJ_BYTE* buffer) {
+	OPJ_INT32 i = 0;
+	OPJ_SIZE_T offset = 0;
+
+	if (!buffer || !seg_buf)
+		return OPJ_FALSE;
+
+	for (i = 0; i < seg_buf->segments.size; ++i) {
+		opj_buffer_t* seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, i);
+		if (seg->len)
+			memcpy(buffer + offset, seg->buf, seg->len);
+		offset += seg->len;
+	}
+	return OPJ_TRUE;
+
+}
+
+OPJ_BYTE* opj_seg_buf_get_global_ptr(opj_seg_buf_t* seg_buf) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return NULL;
+	cur_seg = (opj_buffer_t*)(seg_buf->segments.data[seg_buf->cur_seg_id]);
+	return (cur_seg) ? (cur_seg->buf + cur_seg->offset) : NULL;
+}
+
+OPJ_SIZE_T opj_seg_buf_get_cur_seg_len(opj_seg_buf_t* seg_buf) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return 0;
+	cur_seg = (opj_buffer_t*)(opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id));
+	return (cur_seg) ? (OPJ_SIZE_T)(cur_seg->len - cur_seg->offset) : 0;
+}
+
+OPJ_OFF_T opj_seg_buf_get_cur_seg_offset(opj_seg_buf_t* seg_buf) {
+	opj_buffer_t* cur_seg = NULL;
+	if (!seg_buf)
+		return 0;
+	cur_seg = (opj_buffer_t*)(opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id));
+	return (cur_seg) ? (OPJ_OFF_T)(cur_seg->offset) : 0;
+}
+
+
+OPJ_OFF_T opj_seg_buf_get_global_offset(opj_seg_buf_t* seg_buf) {
+	OPJ_INT32 i = 0;
+	OPJ_OFF_T offset = 0;
+
+	if (!seg_buf)
+		return 0;
+
+	for (i = 0; i < seg_buf->cur_seg_id; ++i) {
+		opj_buffer_t* seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, i);
+		offset += seg->len;
+	}
+	return offset + opj_seg_buf_get_cur_seg_offset(seg_buf);
+}
+
+
+void opj_buffer_incr_offset(opj_buffer_t* buf, OPJ_OFF_T off) {
+	if (!buf)
+		return;
+	/*  we allow the offset to move to one location beyond end of buffer segment*/
+	if (buf->offset + off > (OPJ_OFF_T)buf->len) {
+		printf("Warning: attempt to increment buffer offset out of bounds\n");
```

Comment:
There should be no printf in the library code. Please use the existing message mechanism if necessary.